Sources/SwiftGRPC/Core/Metadata.swift

```diff
@@ -20,25 +20,42 @@ import Foundation // for String.Encoding
 
 /// Metadata sent with gRPC messages
 public class Metadata: CustomStringConvertible {
+  public enum Error: Swift.Error {
+    /// Field ownership can only be transferred once. Likewise, it is not advisable to write to a metadata array whose
+    /// fields we do not own.
+    case doesNotOwnFields
+  }
+  
   /// Pointer to underlying C representation
-  let underlyingArray: UnsafeMutableRawPointer
+  fileprivate let underlyingArray: UnsafeMutableRawPointer
+  /// Ownership of the fields inside metadata arrays provided by `grpc_op_recv_initial_metadata` and
+  /// `grpc_op_recv_status_on_client` is retained by the gRPC library. Similarly, passing metadata to gRPC for sending
+  /// to the client for sending/receiving also transfers ownership. However, before we have passed that metadata to
+  /// gRPC, we are still responsible for releasing its fields. This variable tracks that.
+  fileprivate var ownsFields: Bool
 
-  init(underlyingArray: UnsafeMutableRawPointer) {
+  init(underlyingArray: UnsafeMutableRawPointer, ownsFields: Bool) {
     self.underlyingArray = underlyingArray
+    self.ownsFields = ownsFields
   }
 
   public init() {
     underlyingArray = cgrpc_metadata_array_create()
+    ownsFields = true
   }
 
-  public init(_ pairs: [String: String]) {
+  public init(_ pairs: [String: String]) throws {
```

Comment:
If I read this correctly, it seems that init throws only because it calls add(), which throws an error that will never occur in practice (because ownsFields is true). Did I miss another throw or do you expect add() to throw other errors in the future? If not, it seems better to use try! add and not throw from init.

Reply:
I had this before (see https://github.com/grpc/grpc-swift/pull/217/commits/7b4f47d91dd2980f13c146979760badf43045e37), but decided that it is safer not to rely on `add` not failing in that context. It is guaranteed to not fail _now_, but that might change if the implementation of `add` gets changes. And I think most clients won't have a problem adding the `try` statement.

Reply:
SGTM.